File: 152-maximum-product-subarray/maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        #  just because the multiplication operation behaves differently with the negative numbers we need to keep track of the curreMin and currMax values

        maxPro = max(nums)
        currMin, currMax = 1, 1

        for n in nums:
            if(n==0):
                currMin, currMax = 1, 1
                continue
            
            temp = currMax * n
            currMax = max(n*currMin, n*currMax, n)
            currMin = min(n*currMin, temp, n)
            maxPro = max(maxPro, currMax)

        return maxPro

File: 152-maximum-product-subarray/maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {

        // Initialize variables
        int currMax = nums[0];
        int currMin = nums[0];
        int maxPro = nums[0]; // Start with the first element as the initial maximum

        // Traverse the array from the second element
        for (int i = 1; i < nums.length; i++) {
            int num = nums[i];

            // Swap currMax and currMin if num is negative
            if (num < 0) {
                int temp = currMax;
                currMax = currMin;
                currMin = temp;
            }

            // Update currMax and currMin
            currMax = Math.max(num, currMax * num);
            currMin = Math.min(num, currMin * num);

            // Update maxPro to store the maximum product found so far
            maxPro = Math.max(maxPro, currMax);
        }

        return maxPro;
    }
}